src/gt4py/next/embedded/nd_array_field.py

```diff
@@ -194,6 +196,113 @@ def restrict(self: _BaseNdArrayField, domain) -> _BaseNdArrayField:
+                and isinstance(idx[0], Dimension)
+                and isinstance(idx[1], UnitRange)
+                or isinstance(idx[1], int)
+                for idx in index
+            ):
+                return self._getitem_absolute_slice(index)
+        elif isinstance(index, Domain):
+            return self._getitem_absolute_slice(index)
+
+        raise IndexError(f"Unsupported index type: {index}")
+
+    def _getitem_absolute_slice(self, index: DomainSlice) -> common.Field:
+        slices = _get_slices_from_domain_slice(self.domain, index)
+        new = self.ndarray[slices]
+
+        # handle single value return
+        if len(new.shape) == 0:
+            return new
+
+        # construct domain
+        if all(isinstance(idx[0], Dimension) and isinstance(idx[1], UnitRange) for idx in index):
+            dims, ranges = zip(*index)
+            new_domain = common.Domain(dims=dims, ranges=ranges)
+        elif all(isinstance(idx[0], Dimension) and isinstance(idx[1], int) for idx in index):
+            new_dims = (self.domain.dims[len(slices) - 1],)
+            new_ranges = (self.domain.ranges[len(slices) - 1],)
+            new_domain = common.Domain(dims=new_dims, ranges=new_ranges)
+
+        return common.field(new, domain=new_domain)
+
+    def _getitem_relative_slice(self, index: tuple[slice | int, ...]) -> common.Field:
+        new = self.ndarray[index]
+        new_dims = []
+        new_ranges = []
+
+        dim_diff = len(new.shape) - len(index)
+
+        if dim_diff > 0:
+            new_index = tuple([*index, Ellipsis])
+        else:
+            new_index = index
+
+        for i, elem in enumerate(new_index):
+            if isinstance(elem, slice):
+                new_dims.append(self.domain.dims[i])
+                new_ranges.append(self._slice_range(self.domain.ranges[i], elem))
+            elif isinstance(elem, int):
+                new_dims.append(self.domain.dims[elem])
+                new_ranges.append(self.domain.ranges[elem])
+            elif isinstance(elem, type(Ellipsis)):
+                new_dims.append(self.domain.dims[i])
+                new_ranges.append(self.domain.ranges[i])
+
+        new_domain = Domain(dims=new_dims, ranges=tuple(new_ranges))
+        return common.field(new, domain=new_domain)
+
+    def _slice_range(self, input_range: UnitRange, slice_obj: slice) -> UnitRange:
+        if slice_obj.start is None:
+            slice_start = 0
+        else:
+            slice_start = (
+                slice_obj.start if slice_obj.start >= 0 else input_range.stop + slice_obj.start
+            )
+
+        if slice_obj.stop is None:
+            slice_stop = 0
+        else:
+            slice_stop = (
+                slice_obj.stop if slice_obj.stop >= 0 else input_range.stop + slice_obj.stop
+            )
+
+        start = input_range.start + slice_start
+        stop = input_range.start + slice_stop
+
+        return UnitRange(start, stop)
```

Comment:
testcase:
input_range = UnitRange(2,10)
slice_obj = slice(2,-2)
expected = UnitRange(4,8)

result:
slice_start = 2
slice_stop = 10 + -2 = 8
start = 2 + 2 = 4 (correct)
stop = 2 + 8 = 10 (wrong)

if I executed the code correctly. If in doubt, test this function separately.

Maybe
```suggestion
        start = (input_range.start if (slice_obj.start or 0) >= 0 else input_range.stop) + slice_obj.start
        stop = (input_range.start if (slice_obj.stop or len(input_range)) >= 0 else input_range.stop) + slice_obj.stop
        return UnitRange(start, stop)
```

Reply:
You marked this as resolved, but it still looks wrong to me. Did you test my example?

Reply:
Should be fixed now